wpilibj/src/main/java/edu/wpi/first/wpilibj/experimental/controller/PIDController.java

```diff
@@ -0,0 +1,581 @@
+   * @param toleranceType Whether the given tolerance values are absolute, or percentages of the
+   *                      total input range.
+   * @return Whether the error is within the acceptable bounds.
+   */
+  public boolean atReference(double tolerance, double deltaTolerance, Tolerance toleranceType) {
+    double error = getError();
+
+    m_thisMutex.lock();
+    try {
+      double deltaError = (error - m_prevError) / getPeriod();
+      if (toleranceType == Tolerance.kPercent) {
+        return Math.abs(error) < tolerance / 100 * m_inputRange
+            && Math.abs(deltaError) < deltaTolerance / 100 * m_inputRange;
+      } else {
+        return Math.abs(error) < tolerance
+            && Math.abs(deltaError) < deltaTolerance;
+      }
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the PID controller to consider the input to be continuous,
+   *
+   * <p>Rather then using the max and min input range as constraints, it considers them to be the
+   * same point and automatically calculates the shortest route to the setpoint.
+   */
+  public void setContinuous() {
+    setContinuous(true);
+  }
+
+  /**
+   * Set the PID controller to consider the input to be continuous,
+   *
+   * <p>Rather then using the max and min input range as constraints, it considers them to be the
+   * same point and automatically calculates the shortest route to the setpoint.
+   *
+   * @param continuous true turns on continuous, false turns off continuous
+   */
+  public void setContinuous(boolean continuous) {
+    m_thisMutex.lock();
+    try {
+      m_continuous = continuous;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Sets the maximum and minimum values expected from the input.
+   *
+   * @param minimumInput the minimum value expected from the input
+   * @param maximumInput the maximum value expected from the output
+   */
+  public void setInputRange(double minimumInput, double maximumInput) {
+    m_thisMutex.lock();
+    try {
+      m_minimumInput = minimumInput;
+      m_maximumInput = maximumInput;
+      m_inputRange = maximumInput - minimumInput;
+    } finally {
+      m_thisMutex.unlock();
+    }
+
+    setReference(getReference());
+  }
+
+  /**
+   * Sets the minimum and maximum values to write.
+   *
+   * @param minimumOutput the minimum value to write to the output
+   * @param maximumOutput the maximum value to write to the output
+   */
+  public void setOutputRange(double minimumOutput, double maximumOutput) {
+    m_thisMutex.lock();
+    try {
+      m_minimumOutput = minimumOutput;
+      m_maximumOutput = maximumOutput;
+    } finally {
+      m_thisMutex.unlock();
+    }
+  }
+
+  /**
+   * Set the absolute error which is considered tolerable for use with atReference().
+   *
+   * @param tolerance Absolute error which is tolerable.
+   */
+  public void setAbsoluteTolerance(double tolerance) {
+    setAbsoluteTolerance(tolerance, Double.POSITIVE_INFINITY);
```

Comment:
Let's add something to the contract that says the delta tolerance will be infinity after calling this method.